GVFS/GVFS.UnitTests/Common/NuGetUpgraderTests.cs

```diff
@@ -0,0 +1,300 @@
+        private MockTracer tracer;
+
+        private NuGetUpgrader.NugetUpgraderConfig upgraderConfig;
+        private string downloadFolder;
+
+        private Mock<NuGetFeed> mockNuGetFeed;
+        private Mock<PhysicalFileSystem> mockFileSystem;
+
+        [SetUp]
+        public void SetUp()
+        {
+            this.upgraderConfig = new NuGetUpgrader.NugetUpgraderConfig(this.tracer, null, NuGetFeedUrl, NuGetFeedName, NuGetFeedUrlForCredentials);
+            this.downloadFolder = "downloadFolderTestValue";
+
+            this.tracer = new MockTracer();
+
+            this.mockNuGetFeed = new Mock<NuGetFeed>(
+                NuGetFeedUrl,
+                NuGetFeedName,
+                this.downloadFolder,
+                null,
+                this.tracer);
+            this.mockFileSystem = new Mock<PhysicalFileSystem>();
+
+            this.upgrader = new NuGetUpgrader(
+                CurrentVersion,
+                this.tracer,
+                this.upgraderConfig,
+                false,
+                this.mockFileSystem.Object,
+                this.mockNuGetFeed.Object,
+                new LocalUpgraderServices(this.tracer, this.mockFileSystem.Object));
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            this.mockNuGetFeed.Object.Dispose();
+        }
+
+        [TestCase]
+        public void TryQueryNewestVersion_NewVersionAvailable()
+        {
+            Version newVersion;
+            string message;
+            List<IPackageSearchMetadata> availablePackages = new List<IPackageSearchMetadata>()
+            {
+                this.GeneratePackageSeachMetadata(new Version(CurrentVersion)),
+                this.GeneratePackageSeachMetadata(new Version(NewerVersion)),
+            };
+
+            this.mockNuGetFeed.Setup(foo => foo.QueryFeedAsync(It.IsAny<string>())).ReturnsAsync(new List<IPackageSearchMetadata>(availablePackages));
+
+            bool success = this.upgrader.TryQueryNewestVersion(out newVersion, out message);
+
+            // Assert that we found the newer version
+            success.ShouldBeTrue();
+            newVersion.ShouldNotBeNull();
+            newVersion.ShouldEqual<Version>(new Version(NewerVersion));
+            message.ShouldNotBeNull();
+        }
+
+        [TestCase]
+        public void TryQueryNewestVersion_MultipleNewVersionsAvailable()
+        {
+            Version newVersion;
+            string message;
+            List<IPackageSearchMetadata> availablePackages = new List<IPackageSearchMetadata>()
+            {
+                this.GeneratePackageSeachMetadata(new Version(CurrentVersion)),
+                this.GeneratePackageSeachMetadata(new Version(NewerVersion)),
+                this.GeneratePackageSeachMetadata(new Version(NewerVersion2)),
+            };
+
+            this.mockNuGetFeed.Setup(foo => foo.QueryFeedAsync(It.IsAny<string>())).ReturnsAsync(new List<IPackageSearchMetadata>(availablePackages));
```

Comment:
Why are we essentially returning a `new List<..>(new List<..> { ..elements..}))`?

Could just do

```suggestion
            this.mockNuGetFeed.Setup(foo => foo.QueryFeedAsync(It.IsAny<string>())).ReturnsAsync(availablePackages);
```

Same comment applies elsewhere too.